move staking test spec to node/chainspec

4meta5 opened the pull request:
### What does it do?

Puts the genesis config information used in the polkadot-launch staking tests in `node/test_spec.rs`

This will replace the `staking-test-spec.json`, which required updating the system wasm code field every time anything in the runtime changes.

To run, need to compile the node with flag `--features=test-spec`. This ensures that the staking-test-spec hard coded into `node/test_spec.rs` is only included in the binary when this flag is passed (and is NOT included when the flag is not passed). @joelamouche this will need to be included in the CI

### What important points reviewers should know?

Any long-living genesis config benefits from being coded in rust instead of the json spec because it is easier to maintain (because the json spec requires changing the wasm blob every time there is any runtime change).

### Is there something left for follow-up PRs?

### What alternative implementations were considered?

### Are there relevant PRs or issues in other repositories (Substrate, Polkadot, Frontier, Cumulus)?

### What value does it bring to the blockchain users?

## Checklist

- Does it require a purge of the network?
- You bumped the runtime version if there are breaking changes in the **runtime** ?
- Does it require changes in documentation/tutorials ?

4meta5 commented on tools/yarn.lock:
```diff
@@ -3977,7 +3977,7 @@ pkg@^4.4.9:
 polkadot-launch@PureStake/polkadot-launch#moonbeam-launch:
   version "1.0.12"
-  resolved "https://codeload.github.com/PureStake/polkadot-launch/tar.gz/7cf76daa718d01d90b7f661cfd01fa3406a803e1"
+  resolved "https://codeload.github.com/PureStake/polkadot-launch/tar.gz/c4eea0a021cd75afc9f3bbc9e4a580e473e5e3ed"
```
I don't know why this changed...

joelamouche commented:
>It seems to be failing when adding a new candidate

It's because the first block now takes 70 seconds.
Look at the updated file: https://github.com/PureStake/moonbeam/blob/2961d51cf98b57e2a100bd4845d75d9225c95120/tools/test-staking.ts#L95

4meta5 commented:
so do we need to increase it because that code reflects the state of this branch?

Also, when I run it now it is stalling after starting the first collator
```
Starting a Collator for parachain 1000: 5Ec4AhPZk8STuex8Wsi9TwDtJQxKqzPJRCH7348Xtcs9vZLJ, Collator port : 36335 wsPort : 36946
Added --chain=staking
Added --no-telemetry,--no-prometheus,--author-id=6be02d1d3665660d22ff9624b7be0551ee1ac91b to parachain
Added --execution=wasm to collator
```

joelamouche commented:
I think it's already merged in master. When I checked on Monday, the first block duration had increased to 2 sessions

4meta5 commented:
So it was at 8000 (see the first error where it prints `Waiting 80 seconds`) but now I increase it to 10000 and I am still getting the same error:

```
New Block: #0: undefined, time since last block: 0.01 sec
Waiting 100 seconds
Current status is Ready
Current status is Broadcast
Trace: Error: new candidate should have been added
    at assert (/Users/4meta5/ong/ps/moonbeam/tools/test-staking.ts:18:25)
    at test (/Users/4meta5/ong/ps/moonbeam/tools/test-staking.ts:97:3)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at process.<anonymous> (/Users/4meta5/ong/ps/moonbeam/tools/node_modules/polkadot-launch/dist/src/index.js:186:17)
    at process.emit (events.js:326:22)
    at process.EventEmitter.emit (domain.js:486:12)
    at process.emit (/Users/4meta5/ong/ps/moonbeam/tools/node_modules/source-map-support/source-map-support.js:495:21)
    at processPromiseRejections (internal/process/promises.js:245:33)
    at processTicksAndRejections (internal/process/task_queues.js:94:32)
Trace: Error: new candidate should have been added
    at assert (/Users/4meta5/ong/ps/moonbeam/tools/test-staking.ts:18:25)
    at test (/Users/4meta5/ong/ps/moonbeam/tools/test-staking.ts:97:3)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at process.<anonymous> (/Users/4meta5/ong/ps/moonbeam/tools/node_modules/polkadot-launch/dist/scripts_moonbeam/test_multi_node_transfers.js:159:17)
    at process.emit (events.js:326:22)
    at process.EventEmitter.emit (domain.js:486:12)
    at process.emit (/Users/4meta5/ong/ps/moonbeam/tools/node_modules/source-map-support/source-map-support.js:495:21)
    at processPromiseRejections (internal/process/promises.js:245:33)
    at processTicksAndRejections (internal/process/task_queues.js:94:32)
```

4meta5 commented:
TODO: use promises like XCMP tests so that `wait` is unnecessary, need to wait until the call succeeds to check state after

joelamouche commented:
>use promises like XCMP tests so that wait is unnecessary, need to wait until the call succeeds to check state after

The reason why this wouldnt work before is because we need to use the inBlock event instead of isFinalized I think. If you create an issue for that, I would be happy to take care of it.

>So it was at 80000 (see the first error where it prints Waiting 80 seconds) but now I increase it to 100000 and I am still getting the same error:

From the logs, it looks like the block #1 still hasnt been created when the error fires

JoshOrndorff reviewed (commented):
Spec-wise this is exactly what I was suggesting :)

4meta5 commented:
Tests pass now, need to hide behind feature flag.

joelamouche commented:
I'm getting `Error: Input("Error opening spec file: No such file or directory (os error 2)")` on the collator

joelamouche commented:
ha ok its probably because I didn't run `--features=test-staking`
But we don't want to have to rebuild it for every test do we?

joelamouche commented on node/src/chain_spec.rs:
```diff
@@ -203,6 +203,52 @@ pub fn get_chain_spec(para_id: ParaId) -> ChainSpec {
+		"staking",
+		ChainType::Local,
+		move || {
+			testnet_genesis(
```
Would be nice to have a comment explaining what each line is for so that I can modify it myself. Does this allow changing the config of every pallet we have, like the specs did?

4meta5 replied:
It's calling the function a few lines below called `testnet_genesis` so you can't understand the config without looking at what that function does with it. I added some comments to improve readability though.

4meta5 replied:
It provides _almost_ as much configurability as the json spec. Instead of setting the balances for accounts, it mints `1 << 80` balance for the specified `endowed_accounts: Vec<AccountId>`.

JoshOrndorff replied:
> It provides almost as much configurability

If the current approach doesn't suit your needs, we can change that `testnet_genesis` function or remove it, or whatever makes it most expressive.

joelamouche reviewed (commented):
<img width="531" alt="Screenshot 2021-04-09 at 16 48 07" src="https://user-images.githubusercontent.com/16758383/114214084-2adf7900-9964-11eb-8e63-78ddcf1c18de.png">
Block 1 won't get produced :'( What did I miss?

4meta5 commented:
> But we don't want to have to rebuild it for every test do we?

We only have to build it once. The binary compiled with the feature flag will work with the other tests because it only adds the `staking-spec` genesis; it doesn't remove or change anything else.

> Block 1 won't get produced :'( What did I miss? 

Here is the output from when I ran it. It took ~80 seconds to produce the first block:

```
Registering Parachain 1000
--- Submitting extrinsic to register parachain 1000. (nonce: 0) ---
Current registration status is Ready
Current registration status is {"broadcast":["12D3KooWRLWAYAeHbtNWDqcUxER6k3t29GhiST2GtjYqRhugW3Fd","12D3KooWFfFbigRAadcRmXJrZo2He6Ccj4PTnAJXN2eeV6wKKpvz","12D3KooWL7AenSfh7cpFVp1BnHZHZ95QeGwAF7UjbXQvDkd5Vcfo","12D3KooWLYcDaeXWjBhrciqt2zJyhDuyiT8Ge6EDGBvV6KuAFeJK"]}
Current registration status is {"inBlock":"0x6b83b9ff5d93aa2bb205af2a084b0c0bb9bd66f0d4966528114380c963a177c7"}
Transaction included at blockHash 0x6b83b9ff5d93aa2bb205af2a084b0c0bb9bd66f0d4966528114380c963a177c7
one REGISTRATION DONE
--- Submitting extrinsic to set balance of 5Ec4AhPZk8STuex8Wsi9TwDtJQxKqzPJRCH7348Xtcs9vZLJ to 1000. (nonce: 1) ---
setbalance finished
Current setBalance status is Ready
Current setBalance status is {"broadcast":["12D3KooWRLWAYAeHbtNWDqcUxER6k3t29GhiST2GtjYqRhugW3Fd","12D3KooWFfFbigRAadcRmXJrZo2He6Ccj4PTnAJXN2eeV6wKKpvz","12D3KooWL7AenSfh7cpFVp1BnHZHZ95QeGwAF7UjbXQvDkd5Vcfo","12D3KooWLYcDaeXWjBhrciqt2zJyhDuyiT8Ge6EDGBvV6KuAFeJK"]}
Current setBalance status is {"inBlock":"0x0f71c8ba9f867570196a7b3e8821ab1c8348fd6ccd9eb3868066a3058342f7b9"}
Transaction included at blockHash 0x0f71c8ba9f867570196a7b3e8821ab1c8348fd6ccd9eb3868066a3058342f7b9
one REGISTRATION DONE
Starting a Collator for parachain 1000: 5Ec4AhPZk8STuex8Wsi9TwDtJQxKqzPJRCH7348Xtcs9vZLJ, Collator port : 36336 wsPort : 36947
Added --chain=staking
Added --no-telemetry,--no-prometheus,--author-id=C0F0f4ab324C46e55D02D0033343B4Be8A55532d to parachain
Added --execution=wasm to collator
Starting a Collator for parachain 1000: 5Ec4AhPZk8STuex8Wsi9TwDtJQxKqzPJRCH7348Xtcs9vZLJ, Collator port : 36337 wsPort : 36948
Added --chain=staking
Added --no-telemetry,--no-prometheus,--author-id=Ff64d3F6efE2317EE2807d223a0Bdc4c0c49dfDB to parachain
Added --execution=wasm to collator
ALL PARACHAINS REGISTERED
New Block: #0: undefined, time since last block: 0.008 sec
Current status is Ready
Current status is Broadcast
New Block: #1: 0x6Be02d1d3665660d22FF9624b7BE0551ee1Ac91b, time since last block: 110.947 sec
Current status is InBlock
Transaction included in Block at blockHash 0x980dd12a28410ca5944cf6840d41cfbae10b8f5922315b1ba24b6dc112408edd
	' {"applyExtrinsic":3}: balances.Reserved:: ["0xFf64d3F6efE2317EE2807d223a0Bdc4c0c49dfDB","0x000000000000003635c9adc5dea00000"]
	' {"applyExtrinsic":3}: parachainStaking.JoinedCollatorCandidates:: ["0xFf64d3F6efE2317EE2807d223a0Bdc4c0c49dfDB","0x000000000000003635c9adc5dea00000","0x00000000000000a2a15d09519be00000"]
	' {"applyExtrinsic":3}: system.ExtrinsicSuccess:: [{"weight":0,"class":"Normal","paysFee":"Yes"}]
Current status is Ready
Current status is Broadcast
New Block: #2: 0xC0F0f4ab324C46e55D02D0033343B4Be8A55532d, time since last block: 12.003 sec
New Block: #3: 0x6Be02d1d3665660d22FF9624b7BE0551ee1Ac91b, time since last block: 11.988 sec
Current status is InBlock
Transaction included in Block at blockHash 0xe1a4c08b02c79f23566b5fea7f006208e9bd085dc81fdcff23d752396447fe5f
	' {"applyExtrinsic":3}: balances.Reserved:: ["0xFf64d3F6efE2317EE2807d223a0Bdc4c0c49dfDB","0x000000000000003635c9adc5dea00000"]
	' {"applyExtrinsic":3}: parachainStaking.CollatorBondedMore:: ["0xFf64d3F6efE2317EE2807d223a0Bdc4c0c49dfDB","0x000000000000003635c9adc5dea00000","0x000000000000006c6b935b8bbd400000"]
	' {"applyExtrinsic":3}: system.ExtrinsicSuccess:: [{"weight":0,"class":"Normal","paysFee":"Yes"}]
Current status is Ready
Current status is Broadcast
New Block: #4: 0xC0F0f4ab324C46e55D02D0033343B4Be8A55532d, time since last block: 12.002 sec
New Block: #5: 0xC0F0f4ab324C46e55D02D0033343B4Be8A55532d, time since last block: 12.035 sec
Current status is InBlock
Transaction included in Block at blockHash 0x12fcf7288368a83f5d8e1d7d0801267232f1db95d979810b91d4c98871da97ad
	' {"applyExtrinsic":3}: balances.Unreserved:: ["0xFf64d3F6efE2317EE2807d223a0Bdc4c0c49dfDB","0x000000000000003635c9adc5dea00000"]
	' {"applyExtrinsic":3}: parachainStaking.CollatorBondedLess:: ["0xFf64d3F6efE2317EE2807d223a0Bdc4c0c49dfDB","0x000000000000006c6b935b8bbd400000","0x000000000000003635c9adc5dea00000"]
	' {"applyExtrinsic":3}: system.ExtrinsicSuccess:: [{"weight":0,"class":"Normal","paysFee":"Yes"}]
Current status is Ready
Current status is Broadcast
New Block: #6: 0xC0F0f4ab324C46e55D02D0033343B4Be8A55532d, time since last block: 11.973 sec
New Block: #7: 0x6Be02d1d3665660d22FF9624b7BE0551ee1Ac91b, time since last block: 11.996 sec
Current status is InBlock
Transaction included in Block at blockHash 0x7465cd8891c9300261d740a35ec5da77468fc91165f0542a9775109ed09e82e4
	' {"applyExtrinsic":3}: balances.Reserved:: ["0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac","0x00000000000000004563918244f40000"]
	' {"applyExtrinsic":3}: parachainStaking.Nomination:: ["0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac","0x00000000000000004563918244f40000","0x6Be02d1d3665660d22FF9624b7BE0551ee1Ac91b","0x00000000000000367b2d3f4823940000"]
	' {"applyExtrinsic":3}: system.ExtrinsicSuccess:: [{"weight":0,"class":"Normal","paysFee":"Yes"}]
Current status is Ready
Current status is Broadcast
New Block: #8: 0x6Be02d1d3665660d22FF9624b7BE0551ee1Ac91b, time since last block: 11.999 sec
New Block: #9: 0x6Be02d1d3665660d22FF9624b7BE0551ee1Ac91b, time since last block: 12.001 sec
Current status is InBlock
Transaction included in Block at blockHash 0x9798a2eedc08203b977b8600efb67e05969f34cbcc1731f46493c48e74b815fa
	' {"applyExtrinsic":3}: balances.Unreserved:: ["0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac","0x00000000000000004563918244f40000"]
	' {"applyExtrinsic":3}: parachainStaking.NominatorLeftCollator:: ["0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac","0x6Be02d1d3665660d22FF9624b7BE0551ee1Ac91b","0x00000000000000004563918244f40000","0x000000000000003635c9adc5dea00000"]
	' {"applyExtrinsic":3}: parachainStaking.NominatorLeft:: ["0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac","0x00000000000000004563918244f40000"]
	' {"applyExtrinsic":3}: system.ExtrinsicSuccess:: [{"weight":0,"class":"Normal","paysFee":"Yes"}]
SUCCESS
```

JoshOrndorff commented on node/src/command.rs:
```diff
@@ -53,6 +53,8 @@ fn load_spec(
 		)?)),
 		"dev" | "development" => Ok(Box::new(chain_spec::development_chain_spec(None, None))),
 		"local" => Ok(Box::new(chain_spec::get_chain_spec(para_id))),
+		#[cfg(feature = "test-staking")]
+		"staking" => Ok(Box::new(chain_spec::staking_spec(para_id))),
```
This is a great idea. Might make sense to do the same thing for stagenet in the future.

4meta5 commented:
- [ ] move special test-spec to its own file so tis easier to maintain, and can add some docs about maintenance
